Fix b/29451090 adding validation and tests for timeout parameters

mpaggi opened the pull request:
Adding code and tests to check the values of the following properties:
adaptor.docContentTimeoutSecs
adaptor.docHeaderTimeoutSecs
adaptor.heartbeatTimeoutSecs
In case the value set in config file is invalid, default value is returned and a warning is logged.

mpaggi committed:
Adding validation for timeout properties

mpaggi committed:
Adding validation for timeout properties

BrettMichaelJohnson commented on src/com/google/enterprise/adaptor/Config.java:
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
+      boolean allowNeg, boolean allowEmpty) {
+    String secondsAsString = getValue(property).trim();
+    try {
+      if ((null == secondsAsString) || "0".equals(secondsAsString) ||
```
Config.getValue() does not return null.

PapaJoltOmega commented on src/com/google/enterprise/adaptor/Config.java:
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
   }
 
+  public long validateTimeout(String property, String defaultValue,
+      boolean allowNeg, boolean allowEmpty) {
```
Why do we have these last two booleans if they are always passed in false?

PapaJoltOmega commented on src/com/google/enterprise/adaptor/Config.java:
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
+      log.log(Level.WARNING, "Invalid value for " + property
+              + ", it has been set to default " + defaultValue + " seconds.");
+    }
+    // For invalid value, returning default
```
Why do we decide to continue with default rather than die?

BrettMichaelJohnson replied:
I agree with PJ.  For truly invalid values, throw InvalidConfigurationException.  This also avoids the problems with the defaults below. Although the heartbeatTimeout will still have to fall back to headerTimeout gracefully.

PapaJoltOmega reviewed (commented):
Thank you

BrettMichaelJohnson commented on src/com/google/enterprise/adaptor/Config.java:
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
+        log.log(Level.WARNING, "Invalid value for " + property
+                + ", it has been set to default " + defaultValue + " seconds.");
+      } else {
+          return Long.parseLong(secondsAsString) * 1000;
```
The pedantic mission-to-mars programmer would ensure that multiplying by 1000 will not overflow.

BrettMichaelJohnson replied:
In fact, I would probably have passed in min and max allowed values (which also does away with allowNeg, which is never true anyway).

BrettMichaelJohnson commented on src/com/google/enterprise/adaptor/Config.java:
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
+
   long getAdaptorDocHeaderTimeoutMillis() {
-    return Long.parseLong(getValue("adaptor.docHeaderTimeoutSecs")) * 1000;
+    return validateTimeout("adaptor.docHeaderTimeoutSecs", "30", false, false);
```
You are restating the default values supplied in the constructor here. I don't like that for 2 reasons:  
1) If the defaults are changed in the future, the change may be missed here.  
2) The default value may have been overridden by the connector using overrideKey(). (The FS adaptor overrides these values)

One solution might be to pull the default values from defaultConfig, but you need to be wary of nulls and computed values (not the case for these).

BrettMichaelJohnson commented on src/com/google/enterprise/adaptor/Config.java:
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
     return Long.parseLong(getValue("adaptor.incrementalPollPeriodSecs")) * 1000;
```
you might as well validate this one too.

BrettMichaelJohnson commented on test/com/google/enterprise/adaptor/ConfigTest.java:
```diff
@@ -330,4 +330,48 @@ public void testPropertiesParseEscapeASlash6() throws Exception {
+    for (int  i = 0; i < invalidValuesToVerify.length; i++) {
+      configFile.setFileContents(putInConfig + invalidValuesToVerify[i]);
+      config.load(configFile);
+      assertEquals(Long.parseLong("30000"),
```
Consider using TimeUnit.SECONDS.toMillis(30).

BrettMichaelJohnson commented on src/com/google/enterprise/adaptor/Config.java:
```diff
@@ -643,20 +643,40 @@ long getAdaptorIncrementalPollPeriodMillis() {
-    return Long.parseLong(secondsAsString) * 1000;
+    return validateTimeout("adaptor.heartbeatTimeoutSecs", Long.toString(
+                           getAdaptorDocHeaderTimeoutMillis() / 1000),
+                           false, false);
```
The default value of heartbeatTimeoutSecs is "", so allowEmpty should probably be true. But that will still generate a WARNING in the logs.

BrettMichaelJohnson commented on test/com/google/enterprise/adaptor/ConfigTest.java:
```diff
@@ -330,4 +330,48 @@ public void testPropertiesParseEscapeASlash6() throws Exception {
     config.load(configFile);
     assertEquals("\\\t\\\f", config.getValue("slash"));
   }
+
```
Add tests that verify that valid timeout config values are returned.

BrettMichaelJohnson commented:
Michele, we are planning a release of the adaptor library this week.  Please consider finishing this up so that is may be included.

PapaJoltOmega commented:
Michele, would you mind if Brett took this CL over? Thank you

mpaggi committed:
Add files via upload

mpaggi committed:
Adding validation for timeout properties

mpaggi committed:
Adding validation for timeout properties

mpaggi committed:
Adding validation for timeout properties

mpaggi commented:
Sorry I did not have time to work on this before. Here the changes suggested.
I changed also the tests since 'thrown.expect' does not seems to works in a for cycle.
No problem for me if Brett wants to take over this CL.

BrettMichaelJohnson commented on src/com/google/enterprise/adaptor/Config.java:
```diff
@@ -639,24 +639,42 @@ String getAdaptorFullListingSchedule() {
 
+  public long validateTimeout(String property) {
+    String secondsAsString = getValue(property).trim();
+    if ("adaptor.heartbeatTimeoutSecs".equalsIgnoreCase(property) &&
```
This belongs in getAdaptorHeartbeatTimeoutMillis, not here.

BrettMichaelJohnson commented on test/com/google/enterprise/adaptor/ConfigTest.java:
```diff
@@ -330,4 +332,107 @@ public void testPropertiesParseEscapeASlash6() throws Exception {
+    String putInConfig = " gsa.hostname=not_used\n"
+                         + "adaptor.incrementalPollPeriodSecs=";
+    String invalidValuesToVerify[] = {"0", "-15", "NotValidValue", "",
+                                      Long.toString(Long.MAX_VALUE + 1)};
```
Long.MAX_VALUE + 1 = Long.MIN_VALUE
What you want is (Long.MAX_VALUE/1000)+1 , which will overflow when converted to milliseconds.

BrettMichaelJohnson commented on test/com/google/enterprise/adaptor/ConfigTest.java:
```diff
@@ -330,4 +332,107 @@ public void testPropertiesParseEscapeASlash6() throws Exception {
+                               + "adaptor.incrementalPollPeriodSecs=1700\n"
+                               + "adaptor.docHeaderTimeoutSecs=50\n"
+                               + "adaptor.docContentTimeoutSecs=190\n"
+                               + "adaptor.heartbeatTimeoutSecs=\n");
```
In this test, explicitly set heartbeat, then add another test that verifies the fallback value.

BrettMichaelJohnson committed:
Move check for fallback value of heartbeatTimeoutSecs

BrettMichaelJohnson committed:
Fix Timeout validation tests

Corrected secs-to-millis overflow values.
Added explicit test for no heartbeat timeout defaulting to header timeout.
Fixed assertEquals(expectedValue, actualValue) order.
Style guide whitespace around =.

BrettMichaelJohnson commented:
PJ,
I took over this CL, and made (admittedly my own) suggested changes.
Please review.

PapaJoltOmega commented on src/com/google/enterprise/adaptor/Config.java:
```diff
@@ -639,24 +639,41 @@ String getAdaptorFullListingSchedule() {
+        return Long.parseLong(secondsAsString) * 1000;
+      } catch (NumberFormatException nfe) {
+        throw new InvalidConfigurationException("Invalid value for "
+            + property + ". Only a numeric value is accepted.");
```
Can we say numeric that also conveys integer?

PapaJoltOmega reviewed (approved):
LGTM.  Thank you

BrettMichaelJohnson committed:
Update error message "numeric" -> "positive integer"
